src/triples/generation.rs

```diff
@@ -441,6 +445,578 @@ async fn do_generation<C: CSCurve>(
+        my_phi_proof1v.push(my_phi_proof1);
+    }
+    
+    // Spec 2.7
+    let wait2 = chan.next_waitpoint();
+    {
+        chan.send_many(
+            wait2,
+            &(
+                &big_e_iv,
+                &big_f_iv,
+                &big_l_iv,
+                &my_randomizers,
+                &my_phi_proof0v,
+                &my_phi_proof1v
+            ),
+        )
+        .await;
+    }
+
+    // Spec 2.8
+    let wait3 = chan.next_waitpoint();
+    for p in participants.others(me) {
+        let mut a_i_jv = vec![];
+        let mut b_i_jv = vec![];
+        for i in 0..N {
+            let e = &ev[i];
+            let f = &fv[i];
+            let a_i_j: ScalarPrimitive<C> = e.evaluate(&p.scalar::<C>()).into();
+            let b_i_j: ScalarPrimitive<C> = f.evaluate(&p.scalar::<C>()).into();
+            a_i_jv.push(a_i_j);
+            b_i_jv.push(b_i_j);
+        }
+        chan.send_private(wait3, p, &(a_i_jv, b_i_jv)).await;
+    }
+    let mut a_iv = vec![];
+    let mut b_iv = vec![];
+    for i in 0..N {
+        let e = &ev[i];
+        let f = &fv[i];
+        let a_i = e.evaluate(&me.scalar::<C>());
+        let b_i = f.evaluate(&me.scalar::<C>());
+        a_iv.push(a_i);
+        b_iv.push(b_i);
+    }
+
+    // Spec 3.1 + 3.2
+    let mut seen = ParticipantCounter::new(&participants);
+    seen.put(me);
+    while !seen.full() {
+        let (from, confirmation): (_, Vec<Digest>) = chan.recv(wait1).await?;
+        if !seen.put(from) {
+            continue;
+        }
+        if confirmation != my_confirmations {
+            return Err(ProtocolError::AssertionFailed(format!(
+                "confirmation from {from:?} did not match expectation"
+            )));
+        }
+    }
+    
+    // Spec 3.3 + 3.4, and also part of 3.6, 5.3, for summing up the Es, Fs, and Ls.
+    let mut big_ev = vec![];
+    let mut big_fv = vec![];
+    let mut big_lv = vec![];
+    let mut big_e_j_zerov = vec![];
+    for i in 0..N {
+        big_ev.push(big_e_iv[i].clone());
+        big_fv.push(big_f_iv[i].clone());
+        big_lv.push(big_l_iv[i].clone());
+        big_e_j_zerov.push(ParticipantMap::new(&participants));
+    }
+    seen.clear();
+    seen.put(me);
+    while !seen.full() {
+        let (
+            from,
+            (
+                their_big_ev,
+                their_big_fv,
+                their_big_lv,
+                their_randomizers,
+                their_phi_proof0v,
+                their_phi_proof1v,
+            ),
+        ): (
+            _,
+            (
+                Vec<GroupPolynomial<C>>,
+                Vec<GroupPolynomial<C>>,
+                Vec<GroupPolynomial<C>>,
+                Vec<Randomizer>,
+                Vec<dlog::Proof<C>>,
+                Vec<dlog::Proof<C>>,
+            ),
+        ) = chan.recv(wait2).await?;
+        if !seen.put(from) {
+            continue;
+        }
+        
+        for i in 0..N {
```

Comment:
For the future: seems like there's some possibility of making this loop process the elements in parallel.  Curious what your thoughts are on that possibility someday.

Reply:
Depends on in what sense we mean parallel.

If parallel across cores, I think that the orchestration overhead would defeat the purpose in all cases except extremely large batch sizes-- and it seems like we could get the same benefit in that case by just running multiple large batches in parallel using separate threads. The reason is because if you think about it, somewhat counter intuitively, the need to wait on all threads to finish a unit of work (finish cooperating populating one of these vecs) means we have a lot of dead time where cores are simply waiting on other cores to finish. This can be mitigated somewhat by reframing things into an actor model where each actor continually does work without waiting on other actors, and then using a mailbox to send work between. The code was not arranged this way and rearranging it seemed like a heavier lift that would require an audit.

If parallel at instruction level, I think we can definitely achieve that-- although I would be surprised if Godbolt revealed we weren't already using mmx intrinsics implicitly (the compiler is fairly smart about these dumb loops which is why I wrote them this way). There is obviously room here to explore the possibility of moving some of this generation to a GPU as well.

Reply:
my intuition for this based on having worked vectorizing this library is that it would be practical to generate the beaver triples on GPU via compute shaders. There would need to be multiple passes, and between passes we would need to do network calls to other participants doing the same. The buffer size would be something like a few hundred megabytes worth of triples.

after the change in this PR we will enter a compute bound regime and exit the network bound regime. If we need additional throughput, using compute shaders will give us another order of magnitude improvement in throughput, although it will require us to deploy GPUS onto the nodes.

Reply:
btw if doing via gpu (or via actor approach) the idea would be that each "stage" takes one or more buffers as input and produces one or more buffers as output. The stages would be arranged so that they are all being worked on in parallel (like an assembly line). The fact that we have to do network calls between stages would be irrelevant to throughput and would only be relevant to latency, assuming the pipeline can be kept full.